### PR TITLE
PLT-5478: Hide collapse/expand arrow for unavailable image links

### PR DESCRIPTION
**Pull request by pruthvip**
#### Summary
Problem: Collapse/expand arrow was shown for broken image urls also.
Solution: Whenever an image link is rendered in collapsed setting i.e the preview of the image is not shown after saving the post.
An async call is made to check if the image exist or not.
If the image doesn't exist. linkLoadError is set to true, which hide the collapse/expand arrow.

Note: When image preview is off. Youtube links will have the collapsed/expand arrow. And Image links, will not have the collapsed/expand arrow initially, but once its checked that the image link is valid collapsed/expand arrow will appear

#### Ticket Link
[GitHub](https://github.com/mattermost/platform/issues/7187)
[Jira](https://mattermost.atlassian.net/browse/PLT-5478)

### CODE REVIEW

**Comment by pruthvip** [comment]
hey @esethna . Thanks for testing it out. The logic is that, by default i show arrows for all urls, for which `isLinkImage` returns true. After rendering, I check if the link exist or not. Based on which I hide it. Hence if the a link is a image link which is broken. Then the arrow will come initially and then it will hide. (Assuming the link which u have pasted is broken). 

If we first check if the link exists and then render. It might add some delay and lead to bad user experience. Hence I had kept the check asynchronously.

**Comment by esethna** [comment]
Thanks @pruthvip, I would propose that we instead do the opposite. As in, we post the link without the arrows for all URLs and if the link is a valid image (`isLinkImage` returns true) the arrows should appear. I think this will be the best user experience without causing a delay.

Right now it is a bad and confusing experience to see the arrows and then have then disappear right away. 

Open to suggestions, I appreciate the discussion,

**Comment by pruthvip** [comment]
@esethna I have made the changes. Now when image preview is off. Youtube links will have the collapsed/expand arrow. And Image links, will not have the collapsed/expand arrow initially, but once its checked that the image link is valid  collapsed/expand arrow will appear. Have changed the description of the PR too accordingly.

File: webapp/components/post_view/post_body_additional_content.jsx [hunk 1]
```diff
@@ -45,6 +45,7 @@ export default class PostBodyAdditionalContent extends React.PureComponent {
         this.isLinkToggleable = this.isLinkToggleable.bind(this);
         this.handleLinkLoadError = this.handleLinkLoadError.bind(this);
         this.handleLinkLoaded = this.handleLinkLoaded.bind(this);
+        this.checkIfLinkImageExists = this.checkIfLinkImageExists.bind(this);
```
Comment (saturninoabril) [inline webapp/components/post_view/post_body_additional_content.jsx]: @pruthvip thanks for the contribution! No need to bind this function.

File: webapp/components/post_view/post_body_additional_content.jsx [hunk 2]
```diff
@@ -54,11 +55,19 @@ export default class PostBodyAdditionalContent extends React.PureComponent {
         };
     }
 
+    componentDidMount() {
+        // check the availablity of the image rendered(if any) in the first render.
```
Comment (saturninoabril) [inline webapp/components/post_view/post_body_additional_content.jsx]: typo error: `availablity` should be `availability `

File: webapp/components/post_view/post_body_additional_content.jsx [hunk 3]
```diff
@@ -54,11 +55,19 @@ export default class PostBodyAdditionalContent extends React.PureComponent {
...
+        this.checkIfLinkImageExists();
+    }
+
     componentWillReceiveProps(nextProps) {
         if (nextProps.previewCollapsed !== this.props.previewCollapsed || nextProps.post.message !== this.props.post.message) {
             this.setState({
                 embedVisible: nextProps.previewCollapsed.startsWith('false'),
                 link: Utils.extractFirstLink(nextProps.post.message)
+            }, () => {
+                // check the availablity of the image link
```
Comment (saturninoabril) [inline webapp/components/post_view/post_body_additional_content.jsx]: should be `availability `

File: webapp/components/post_view/post_body_additional_content.jsx [hunk 4]
```diff
@@ -82,6 +91,24 @@ export default class PostBodyAdditionalContent extends React.PureComponent {
         );
     }
 
+    // when image links are collapsed, check if the link is a valid image url and it is available
+    checkIfLinkImageExists() {
```
Comment (saturninoabril) [inline webapp/components/post_view/post_body_additional_content.jsx]: Function name is confusing at first read. Could it be change to something like: `precheckImageLink`?

**Comment by pruthvip** [comment]
Thanks for suggestions @saturninoabril . I have made the changes

**Pull request merged**